### DISCUSSION TITLE
ASReview screening output sorted on labeling time

### DISCUSSION THREAD

**Discussion by JJMFreriksen** [asreview/asreview#1798] [Q&A]
Hi all, 
Is it possible to sort the screening output on time of labeling (add the time of labeling) instead of relevance?
Thanks!

**Comment by PeterLombaers** [comment]
If you know a little bit of Python you can follow [this example notebook](https://asreview.readthedocs.io/en/latest/example_api_asreview_file.html#State-API). It shows how to get the time of labeling for your screening output and how to merge that data with your original input dataset.

**Comment by jteijema** [comment]
```python
from pathlib import Path
from asreview import open_state
from asreview import ASReviewProject
from asreview import ASReviewData
import fire
import tempfile

def get_labeling_times(asreview_file, output_file, sort=False):
    # Load the ASReview dataset
    with open_state(asreview_file) as state:
        df = state.get_dataset()

    # Select relevant columns and set index
    labeling_times = df[['record_id','labeling_time']].set_index("record_id")

    # Create a temporary directory
    temp_dir = tempfile.TemporaryDirectory()
    try:
        project_path = Path(temp_dir.name)
        
        # Load the ASReview project
        project = ASReviewProject.load(asreview_file, project_path)
        dataset_fp = Path(
            project_path, project.config["id"], "data", project.config["dataset_path"]
        )
        dataset = ASReviewData.from_file(dataset_fp).to_dataframe()

        # Merge the dataframes
        merged_df = dataset.join(labeling_times)

        # Sort by labeling_time if the flag is set
        if sort:
            merged_df = merged_df.sort_values('labeling_time')

        # Save to a new Excel file
        merged_df.to_excel(output_file)

        print(f"Output saved to {output_file}")
    
    finally:
        # Cleanup the temporary directory
        temp_dir.cleanup()

if __name__ == '__main__':
    fire.Fire(get_labeling_times)
```

---

## Introduction

This script, get_labeling_times.py, is designed to merge data from an ASReview
project file with the labeling time.
This can be useful for tracking and analyzing the labeling process in systematic
reviews.

## Requirements

To use this script, you need to have the following Python packages installed:

- asreview
- fire
- pandas
- openpyxl

You can install the required packages using pip with the following command:

```bash
pip install asreview fire pandas openpyxl
```

## Usage

Follow these steps to use the script:

1. Download the Script and README:  
    Ensure you have downloaded `get_labeling_times.py` and this README file.

2. Prepare Your ASReview File:  
    Ensure you have an ASReview file exported from the ASReview application. This file should have a `.asreview` extension.

3. Run the Script:  
    Open your command line interface (CLI) and navigate to the directory containing `get_labeling_times.py`.

4. Basic Command:  
    Run the script by using the following command:

    ```bash
    python get_labeling_times.py path_to_asreview_file.asreview path_to_output_xlsx_file.xlsx
    ```
    Replace `path_to_asreview_file.asreview` with the path to your ASReview file and `path_to_output_xlsx_file.xlsx` with the desired path for the output Excel file.

5. Optional Sorting:  
    If you want to sort the data by labeling time, add the `--sort` flag:

    ```bash
    python get_labeling_times.py path_to_asreview_file.asreview path_to_output_xlsx_file.xlsx --sort
    ```

6. Help Command:  
    For more help and information about the script's usage, you can run:

    ```bash
    python get_labeling_times.py --help
    ```

## Example

Here's an example command to run the script:

```bash
python get_labeling_times.py my_asreview_project.asreview output_data.xlsx --sort
```

This command will process the ASReview project file `my_asreview_project.asreview`, merge it with the relevant data, sort the data by labeling time, and save the output to output_data.xlsx.
